src/otel/OTelGroupCallMembership.ts

```diff
@@ -0,0 +1,177 @@
+  flatObject: Record<string, unknown>,
+  prefix: string,
+  depth: number
+) {
+  if (depth > 10)
+    throw new Error(
+      "Depth limit exceeded: aborting VoipEvent recursion. Prefix is " + prefix
+    );
+
+  for (const [k, v] of Object.entries(obj)) {
+    if (["string", "number"].includes(typeof v)) {
+      flatObject[prefix + k] = v;
+    } else if (typeof v === "object") {
+      flattenVoipEventRecursive(
+        v as Record<string, unknown>,
+        flatObject,
+        prefix + k + ".",
+        depth + 1
+      );
+    }
+  }
+}
+
+/**
+ * Represent the span of time which we intend to be joined to a group call
+ */
+export class OTelGroupCallMembership {
+  private callMembershipSpan: Span;
+  private myUserId: string;
+  private myMember: RoomMember;
+
+  constructor(private groupCall: GroupCall, client: MatrixClient) {
+    this.myUserId = client.getUserId();
+    this.myMember = groupCall.room.getMember(client.getUserId());
+
+    provider.resource.attributes[
+      SemanticResourceAttributes.SERVICE_NAME
+    ] = `element-call-${this.myUserId}-${client.getDeviceId()}`;
+  }
+
+  public onJoinCall() {
+    // Create the main span that tracks the time we intend to be in the call
+    this.callMembershipSpan = tracer.startSpan("matrix.groupCallMembership");
+    this.callMembershipSpan.setAttribute(
+      "matrix.confId",
+      this.groupCall.groupCallId
+    );
+    this.callMembershipSpan.setAttribute("matrix.userId", this.myUserId);
+    this.callMembershipSpan.setAttribute(
+      "matrix.displayName",
+      this.myMember.name
+    );
+
+    opentelemetry.trace.setSpan(
+      opentelemetry.context.active(),
+      this.callMembershipSpan
+    );
+
+    this.callMembershipSpan.addEvent("matrix.joinCall");
+  }
+
+  public onLeaveCall() {
+    this.callMembershipSpan.addEvent("matrix.leaveCall");
+
+    // and end the main span to indicate we've left
+    if (this.callMembershipSpan) this.callMembershipSpan.end();
+  }
+
+  public onUpdateRoomState(event: MatrixEvent) {
+    if (
+      !event ||
+      (!event.getType().startsWith("m.call") &&
+        !event.getType().startsWith("org.matrix.msc3401.call"))
+    ) {
+      return;
+    }
+
+    this.callMembershipSpan.addEvent(
+      `otel_onRoomStateEvent_${event.getType()}`,
+      flattenVoipEvent(event.getContent())
+    );
+  }
+
+  public onSendEvent(event: VoipEvent) {
+    const eventType = event.eventType as string;
+    if (!eventType.startsWith("m.call")) return;
+
+    if (event.type === "toDevice") {
+      this.callMembershipSpan.addEvent(
+        `matrix.sendToDeviceEvent_${event.eventType}`,
+        flattenVoipEvent(event)
+      );
+    } else if (event.type === "sendEvent") {
+      this.callMembershipSpan.addEvent(
+        `matrix.sendToRoomEvent_${event.eventType}`,
+        flattenVoipEvent(event)
+      );
+    }
+  }
+
+  public onToggleMicrophoneMuted(newValue: boolean) {
+    this.callMembershipSpan.addEvent("matrix.toggleMicMuted", {
```

Comment:
This errors if you try to mute yourself for the lobby screen - I think `callMembershipSpan` should be nullable